### ISSUE TITLE
Accessing AsyncSnapshot.data should throw if there's an error

### ISSUE THREAD

**Issue by tvolkert** [flutter/flutter#34545]
### Steps to reproduce

Run the following app:

```dart
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  Stream<int> get stream async* {
    throw StateError('error');
  }

  @override
  Widget build(BuildContext context) {
    return Directionality(
      textDirection: TextDirection.ltr,
      child: StreamBuilder<int>(
        stream: stream,
        builder: (BuildContext context, AsyncSnapshot<int> snapshot) {
          return Text('${snapshot.data}');
        },
      ),
    );
  }
}
```

### Expected behavior

You expect the `StateError` to show up in the console logs.

### Actual behavior

The exception is swallowed.

**Comment by tvolkert** [comment]
@Hixie @goderbauer

**Comment by goderbauer** [comment]
My first reaction is: This is working as intended (with a caveat, see below). The StreamBuilder catches exceptions of the stream and feeds them as `AsyncSnapshot`s into the builder. In there it is the developer's responsibility to properly handle the error by checking `snapshot.hasError` and `snapshot.error`. They can then either throw - or present the user with an alternative widget indicating that something went wrong.

We even documented that on the `StreamBuilder`: https://github.com/flutter/flutter/blob/5493fe469e146726d25955326c0bad08af1b1458/packages/flutter/lib/src/widgets/async.dart#L330-L332

Having said all this: For some reason, the builder is currently not called again when the `AsyncSnapshot` changes from `ConnectionState.waiting` to `ConnectionState.active` (with the error condition). I am currently investigating why.

tl;dr: It is intended that nothing throws (Maybe we should document that in a more obvious place to prevent confusion). However, it seems unintended that the builder is not called when the stream moves into the error state.

**Comment by goderbauer** [comment]
See also the sample code on StreamBuilder (line 352/353):

https://github.com/flutter/flutter/blob/5493fe469e146726d25955326c0bad08af1b1458/packages/flutter/lib/src/widgets/async.dart#L349-L361

**Comment by goderbauer** [comment]
Actually, everything is working as intended here:

```dart
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  Stream<int> get stream async* {
    throw StateError('error');
  }

  @override
  Widget build(BuildContext context) {
    return Directionality(
      textDirection: TextDirection.ltr,
      child: StreamBuilder<int>(
        stream: stream,
        builder: (BuildContext context, AsyncSnapshot<int> snapshot) {
          if (snapshot.hasError) {
            return Text('An error has occorued: ${snapshot.error}');
          }
          return Text('${snapshot.data}');
        },
      ),
    );
  }
}
```

**Comment by tvolkert** [comment]
Should `AsyncSnapshot.data` throw if there's an error?

**Comment by goderbauer** [comment]
That does sound like a good idea. It's a small breaking change, but probably worth doing to remind people to check for errors. I'll repurpose this bug.

**Comment by goderbauer** [comment]
Actually, there is already `AsyncSnapshot.requireData`, that basically does this check: https://master-api.flutter.dev/flutter/widgets/AsyncSnapshot/requireData.html

**Comment by tvolkert** [comment]
I wonder if the `requireData` versus `data` distinction is just a foot gun - and if we should remove `requireData` in favor of making `data` doing that job.

**Comment by audkar** [comment]
My 2¢:
- I personally had this problem when started using Flutter. Exceptions were handled silently and we didn't knew that they even happen.
- But... This breaking change AND it will not be visible for developers immediately after Flutter version update in project. For many error will appear only after release at run-time (network errors & stuff). So this would be very sneaky breaking change :)

**Comment by tvolkert** [comment]
We could do a soft transition, whereby only in debug mode (via asserts), we'd throw if the `data` property was accessed without `error` or `hasError` also being consulted.

**Comment by tvolkert** [comment]
Proposed soft-transition period: https://github.com/flutter/flutter/pull/34626

Note that it's still a breaking change because I had to remove the `const` qualifiers from the `AsyncSnapshot` constructors.

**Comment by tvolkert** [comment]
Based on initial discussion in the PR, the most recent idea is to deprecate `AsyncSnapshot.data` altogether in favor of an `AsyncSnapshot.value` property, which will throw if there's an error.  The thought is that it forces all developers to make a change, and in so doing, they'll be curious as to the difference between the old `data` and the new `value`, and upon discovering the difference, they'll update their code accordingly.

I don't love this idea.  `data` is a better name than `value`, as it matches the `onData` callback of stream listeners (and is also reflected in the `initialData` property of `StreamBuilder`).

I'm starting to wonder if we should just make the breaking change.

**Comment by tvolkert** [comment]
Code review has led @goderbauer and I to lean towards just making the breaking change.  Bottom line: callers who are blindly accessing `data` in the current API have broken code, but it's broken in an insidious way: without them being alerted to it via crash reports.

Here's the associated breaking change announcement: https://groups.google.com/forum/#!topic/flutter-announce/H6Od0QdsdrI

I'm updating the PR to match.

**Comment by rrousselGit** [comment]
I disagree with that change.

While this is a step in the right direction, there's a flaw:

`hasData` is just `data != null`, but `null` may be a valid data.


If we're going toward these, I'd recommend updating `hasData` based on whether a value was emitted or not instead of whether it's null or not.

**Comment by rrousselGit** [comment]
A potential fix is to change `hasData`to:

```dart
bool get hasData => !hasError && connectionState != ConnectionState.waiting;
```
(and just check `hasData` in the `data` getter).


It'd then be up to non nullable type to catch the remaining errors.

**Comment by danwulff** [comment]
I agree with @rrousselGit here. We have multiple StreamBuilders where `null` is a valid value. `.hasData` does not seem to behave well with this concept.

What would be the proposed alternative? Checking `.hasError` prior to accessing `.data` instead of `.hasData`?

**Comment by tvolkert** [comment]
@rrousselGit @danwulff yep, that's a good point and use case.

How about if we create a sentinel value that represents "no data yet received", different from null, and throw if data is accessed when we're in that state?

**Comment by rrousselGit** [comment]
That would be `ConnectionState.waiting`/`ConnectionState.none`. But the issue is, there's `initialData` and it can be `null` too.

Maybe just check `hasError` and don't check anything else?

**Comment by goderbauer** [comment]
What do you mean by the `initialData`? `AsyncSnapshot` doesn't seem to have a property for that?

**Comment by rrousselGit** [comment]
Yes but `AsyncSnapshot` is used mainly for `StreamBuilder`/`FutureBuilder`, and they do have one.

```dart
StreamBuilder<T>(
  initialData: t,
  stream: controller.stream,
  builder: (context, snapshot) {
        // `snapshot.hasData` should return `true` even when `snapshot.connectionState` 
       // is `none`/`waiting`
  }
}
```

**Comment by goderbauer** [comment]
In StreamBuilder, the initialData value is used to build an AsyncSnaposhot:

https://github.com/flutter/flutter/blob/5493fe469e146726d25955326c0bad08af1b1458/packages/flutter/lib/src/widgets/async.dart#L403

The `initialData` value (even if it is null) would replace the sentinel value that represents "no data" in the AsyncSnapshot and hence hasData would return true even for the initial data.

**Comment by rrousselGit** [comment]
But then `hasData` will always start `true` for `StreamBuilder`, even if `initialData` is not specified.

We'd have to change [that line](https://github.com/flutter/flutter/blob/5493fe469e146726d25955326c0bad08af1b1458/packages/flutter/lib/src/widgets/async.dart#L403) to:

```dart
AsyncSnapshot<T> initial() => 
  initialData != null
    ? AsyncSnapshot<T>.withData(ConnectionState.none, initialData)
    : AsyncSnapshot<T>.nothing();
```

... which would consider that `null` is not a valid value.

**Comment by goderbauer** [comment]
Instead of changing the API, we could just use another placeholder object there that indicates "no initial data".

**Comment by tvolkert** [comment]
Placeholder values are tricky because these classes are generically typed, so I can't make a sentinel value of type `T`...

We may have to change the initial data API as @rrousselGit proposes.

@goderbauer do you see any alternatives?

**Comment by goderbauer** [comment]
Taking a step back, here is what we know:

* AsyncSnapshot makes it too easy to miss error cases because errors may just get swallowed if you're not careful as a user.
*  null may be valid `data` for a AsyncSnapshot.

About the second bullet: We know based on what constructor was called if null represents data (constructor withData() was called) or if it is a placeholder for no data (constructor nothing() or withError() was used). So instead of defining `hasData` as `data != null` we could just assign a boolean to hasData based on what constructor was used.

The problem now is that StreamBuilder and Co always use the withData() constructor even when initialData is null (and we have no real way of knowing if that null means null as data or represents no data). Maybe we could do the same dual constructor approach here to differentiate between the two: The default constructor wouldn't have an initialData argument and if that one is used we will construct the initial snapshot by calling AsyncSnapshot.nothing(). A second constructor (StreamBuilder.withInitialData) takes a required initialData argument and uses the withData constructor of the Snapshot. If you pass null here, it means null as Data.

That's probably the API we should have designed from the beginning, maybe? What do you think @tvolkert @rrousselGit?

If that sounds good, the question is, can we transition to that API from where we are? The changes to AsyncSnapshot look easy, the only "breaking" change is that the data getter will throw if there is no data, which is the point of the whole exercise. The changes to SteamBuilder and Co. are more breaking. Maybe we can ease the transition by keeping the initialData parameter around on the default constructor and mark it as deprecated for a couple of releases pointing to the new withInitialData constructor? Not sure, if you can even mark individual arguments as deprecated?

**Comment by rrousselGit** [comment]
Splitting the behavior in two constructors is a good idea. It'll help when non nullable types are around (as the presence/absence of initialData has an influence on the prototype of builder).

But I think we can leave the `withInitialData` as the default behavior, and instead of the _without_ variant as a `StreamBuilder.uninitialized`.

Most of the time, we want to specify an initialData. If not for that, then the widget tree will usually flash for a frame because the first value is emitted asynchronously.

**Comment by tvolkert** [comment]
@rrousselGit the thing I like about making the default constructor for `FutureBuilder` and `StreamBuilder` have no data (and providing a new `.withInitialData()` constructor) is that the corresponding breaking change will force users to look for how to update their code -- they'll be passing an argument that's no longer supported, and in searching for how to respond, they'll find the `.withInitialData()` constructor.

Conversely, if we make initial data the default, then callers who currently _aren't_ specifying initial data will see a "warning: The parameter 'initialData' is required." analysis error, and I'm afraid that they'll just say "ok, I guess I'll pass null as the initial data" without looking for or finding the `.withoutInitialData()` constructor.

fwiw, my use cases for `StreamBuilder` in my own app have been with no initial data (e.g. showing a stream of photos by floating them up the display), so it's not clear that the majority use case is to specify initial data.

Thoughts?

**Comment by rrousselGit** [comment]
The "warning that can be fixed with `null`". would be only temporary. It'd transform into a compilation error with non nullable types.

**Comment by tvolkert** [comment]
Not true - we're saying null is a valid value for data (and would remain valid once nnbd lands)

**Comment by rrousselGit** [comment]
Yes but `@required` becomes part of the language and if `T` is non nullable then `initialData` will be too. The prototype would be:

```dart
StreamBuilder<T>({
  required T initialData, 
  required Widget builder(BuildContext, AsyncSnashot<T>)
}
```

Which means:

```dart
StreamBuilder<int>(builder:...) // compile error: initialData is required
StreamBuilder<int>(initialData: null, builder:...) // compile error, nul is not a subtype of int for initialData
StreamBuilder<int>(initialData: 42, builder:...) // valid

StreamBuilder<int?>(builder:...) // compile error: initialData is required
StreamBuilder<int?>(initialData: null, builder:...) // valid
StreamBuilder<int?>(initialData: 42, builder:...) // valid
```

**Comment by tvolkert** [comment]
That's a good point.  Ok, I'll update the PR accordingly.

**Comment by tvolkert** [comment]
After review in this issue and in the PR, here's the current list of breaking changes:

1. Mark the following protected methods in `StreamBuilderBase` as `@protected`.
    1. `initial()`
    1. `afterConnected()`
    1. `afterData()`
    1. `afterError()`
    1. `afterDone()`
    1. `afterDisconnected()`
1. Replace `AsyncSnapshot.nothing()` with `AsyncSnapshot.withoutData(ConnectionState)`.  Affected callers (unlikely) can update their code by calling `AsyncSnapshot.withoutData(ConnectionState.none)`.
1. Make `AsyncSnapshot.data` throw an error if `AsyncSnapshot.hasData` is false.  Callers can guard against this by consulting `AsyncSnapshot.hasData` before accessing the `data` property.
1. Remove `AsyncSnapshot.requireData`.  Callers who are using this can update their code by accessing `AsyncSnapshot.data` instead.
1. Mark the `initialData` argument to the `StreamBuilder()` constructor as `@required`.  Callers who do not wish to pass initial data may use the newly added `StreamBuilder.withoutInitialData` constructor.
1. Deprecate the `initialData` argument to the `FutureBuilder()` constructor.  Callers who wish their async snapshot to have an initial value before the future completes can simply check the connection state of the snapshot in their `AsyncWidgetBuilder`.

**Comment by rrousselGit** [comment]
> Deprecate the initialData argument to the FutureBuilder() constructor. Callers who wish their async snapshot to have an initial value before the future completes can simply check the connection state of the snapshot in their AsyncWidgetBuilder.
 
What's the reasoning behind this change?

**Comment by tvolkert** [comment]
> What's the reasoning behind this change?

There's no reason to have initial data on a FutureBuilder - if you had the initial data synchronously, you would just build the widget with the value immediately, and if you need to know whether the future has completed, just use the connection state.

Basically, the initial data on FutureBuilder provides no value.  e.g. instead of:

```dart
Widget build(BuildContext context) {
  return FutureBuilder<String>(
    future: _calculation,
    initialData: 'initial',
    builder: (BuildContext context, AsyncSnapshot<String> snapshot) {
      return Text(snapshot.data),
    },
  );
}
```

Just do:

```dart
Widget build(BuildContext context) {
  return FutureBuilder<String>(
    future: _calculation,
    builder: (BuildContext context, AsyncSnapshot<String> snapshot) {
      String value = snapshot.connectionState == ConnectionState.done
          ? snapshot.data
          : 'initial';
      return Text(value);
    },
  );
}
```

**Comment by tvolkert** [comment]
@rrousselGit I cc'd you on the PR - I welcome your thoughts on the code review.

**Comment by lucaslcode** [comment]
>     3. Make `AsyncSnapshot.data` throw an error if `AsyncSnapshot.hasData` is false.  Callers can guard against this by consulting `AsyncSnapshot.hasData` before accessing the `data` property.

Some people may be using snapshot.data = null as a loading/error state and the analyzer won't be any help. Searched through some of my code and I did this once or twice when I was very lazy and an error wouldn't matter too much (key for an AnimatedSize).

Easily fixable in my code but also easy to miss. Maybe this makes more sense when there are nullable types?

**Comment by JSANL** [comment]
Sorry if it was already mentioned, but this pattern (which I personally often use for being so concise) would also throw then right?

```dart
return StreamBuilder(
        stream: stream,
        builder: (context, snapshot) {
            return MyListView(
              children: snapshot?.data ?? [],
            )
         }
    );
```

**Comment by tvolkert** [comment]
@lucaslcode 

> Edit: Just saw that it will only throw in debug mode, seems reasonable!

We moved  away from only doing it in debug mode because that introduced other transition problems.

Are you saying that you have code like the following:

```dart
builder: (context, snapshot) {
  if (snapshot.data == null) {
    // interpret this as an error and update the UI accordingly.
  } else {
    // Success.
  }
}
```

In that case, yes, this would now throw (unless your stream actually produced a null value) - and you'll need to update your code to check `snapshot.hasData` rather than checking for null.

@JSANL yes, that case is similar - it should be updated to check `snapshot.hasData`.

The justification for this change is that with the current design, we see several bug reports of people confused that they never saw the stream or future error (the error appeared to have been swallowed).

**Comment by rrousselGit** [comment]
Sorry for the late answer:

> Just do:
> ```dart
> Widget build(BuildContext context) {
>  return FutureBuilder<String>(
>    future: _calculation,
>    builder: (BuildContext context, AsyncSnapshot<String> snapshot) {
>      String value = snapshot.connectionState == ConnectionState.done
>          ? snapshot.data
>          : 'initial';
>      return Text(value);
>    },
>  );
>}
> ```

This is technically different from `initialData`.
`FutureBuilder` also handles the situation where the `Future` instance changes (and/or is `null`).

I don't think this is a common use-case, but `ConnectionState.done` it's worth a mention.

**Comment by tvolkert** [comment]
@rrousselGit sorry, my example was simplistic.  `ConnectionState.none` will represent that the future is null and `ConnectionState.waiting` will represent that the future was non-null but has not yet completed.

**Comment by tvolkert** [comment]
#34626 has been merged, so this is now implemented on the master channel.

**Comment by tvolkert** [comment]
@rrousselGit @danwulff after looking at usages within Google's codebase, I found 34 usages of `StreamBuilder` without initial data, and 12 usages _with_ initial data.

What do you think about making the default `StreamBuilder()` constructor _not_ take initial data, and creating a new `StreamBuilder.withInitialData()` constructor?  To roll this out, I'd start by deprecating the `initialData` argument to the default `StreamBuilder()` constructor and interpreting null initial data as "has no initial data", and use cases where null is a valid value can move to using `StreamBuilder.withInitialData(initialData: null)`.

This is all functionally equivalent to the code that landed in #34626 (but hasn't yet made it to the dev channel), but if Google's codebase is any indication, switching around the constructors in this way would break fewer people.

Thoughts?

**Comment by tvolkert** [comment]
@creativecreatorormaybenot also interested in your thoughts on the comment above.

**Comment by rrousselGit** [comment]
> after looking at usages within Google's codebase, I found 34 usages of StreamBuilder without initial data, and 12 usages with initial data.

Among them, how many of these are broken because of the `.data` change?
Because alternatively to `hasData`, `initialData` is a legitimate fix.

**Comment by creativecreatorormaybenot** [comment]
@tvolkert I found that providing `initialData` now makes the code cleaner, i.e. it makes some parts more concise because there is no need for an additional check and definitely more precise as it is clear if there is a way to provide initial data.

However, looking at my code base, I also used `withoutInitialData` primarily.

On the one hand, it might be strange for people getting into Flutter to use a constructor like `withoutInitialData` by default and on the other hand, it might enforce more aware behavior.  
I think less people will use `withInitialData` if it is the named constructor, however, it might make more sense to use the default constructor for the most common use case.

**Comment by tvolkert** [comment]
Things I like about making the default constructor being without initial data (proposed in https://github.com/flutter/flutter/pull/36577):

1. It allows for a softer transition, since we can now deprecate the `initialData` argument to the default constructor, guiding callers over to the new constructor.  Thus, existing users of `StreamBuilder` (and `FutureBuilder`) will be given a period where their code can work on both Flutter master and stable.  This is super important for plugins and packages on Pub.

1. It has parity with the new `FutureBuilder` constructor, where the `initialData` argument is also deprecated and slated for removal in a future release.

1. It uses a data-based approach to the API design, catering to the more common use case.

**Comment by tvolkert** [comment]
@rrousselGit 

> Among them, how many of these are broken because of the `.data` change?

I just went through them all, and of the 34 usages, 4 are broken by the change (were trying to access `data` without checking `hasData`).

I think given this and https://github.com/flutter/flutter/issues/34545#issuecomment-513408442, I'm going to proceed with #36577

**Comment by rrousselGit** [comment]
But `initialData` is safer.

The current default constructor, that doesn't have one, may generate runtime exceptions. 
While if initialData is the default, then with nbd that'll just make a compilation error.

And similarly, the pre-breaking change behavior is closer to `withInitialData` than without.

It is common for folks to do `snapshot.data?.foo` on stable.

**Comment by tvolkert** [comment]
@rrousselGit that's a good point, as long as callers don't just migrate to `.withoutInitialData()` without updating their `data` checks.

Summarizing, it seems like making initial data the default constructor is more _invasive_ (more of a breaking API) but lands the callers in a safer place.  Conversely, making no initial data the default is less invasive (allows for a softer transition) but by not forcing the transition, it silently puts the callers in a position where they're more likely to fail at runtime.

Does that accurately reflect your concern?

**Comment by rrousselGit** [comment]
Indeed. 
I don't have a strong opinion on that one, but I tend to prefer safety over popularity for the default behavior.

I personally don't see much reasons to _not_ use `initialData` (even for `FutureBuilder`).
`snapshot.data?.foo ?? fallback` is a lot easier to write and still safer than `hasData`.

To have the exact same behavior using `hasData`, we'd have to write:
```dart
snapshot.hasData ? snapshot.data?.foo ?? fallback : fallback
```

In any case, if `initialData` is not on the default constructor, it's totally possible to subclass `StreamBuilder` to make it the default.

**Comment by tvolkert** [comment]
Taking a step back...

Somehow, early on in this discussion, we broadened the scope of the discussion beyond the original problem.  The original problem was that accessing `AsyncSnapshot.data` was not throwing if there was an error (`hasError` is true).

What if we stop conflating `data == null` and initial data with this discussion altogether -- and simply  make the small breaking change of making the `data` getter throw if `hasError` is true?  I would also get rid of `requireData` and let callers check null themselves.

Thoughts?  It's a far less invasive change and solves the originally reported bug!

https://github.com/flutter/flutter/pull/34626 has not yet been rolled to dev...

@goderbauer @rrousselGit @creativecreatorormaybenot @lucaslcode @danwulff

**Comment by tvolkert** [comment]
p.s. the current code on master looks like this when people update:

![stream-builder-default-initial-data](https://user-images.githubusercontent.com/15253456/61584474-21154280-aafd-11e9-8dc9-c7a69a206abb.png)

**Comment by rrousselGit** [comment]
Good idea. 👍

**Comment by goderbauer** [comment]
What was the reason for the revert in https://github.com/flutter/flutter/pull/36618?

> What if we stop conflating data == null and initial data with this discussion altogether -- and simply make the small breaking change of making the data getter throw if hasError is true? I would also get rid of requireData and let callers check null themselves.

That does sound ok to me - with one inconvenience: The conflating of data == null and initial data is kinda backed in to the API in the `hasData` getter (which just checks if data is null, but null may be valid data...). I wonder if we should get rid of hasData as well and just tell people if you want to know if data is not null, just check `data`?

**Comment by creativecreatorormaybenot** [comment]
@tvolkert You might want to adjust [the changelog](https://github.com/flutter/flutter/wiki/Changelog) or did you keep it intentionally?

**Comment by tvolkert** [comment]
@creativecreatorormaybenot thanks - done!

**Comment by ChristofferSkog21st** [comment]
@tvolkert What's the status of this? It's still marked as open but without any apparent activity, I'm finding myself a bit confused as to how to work with the data right now that's why I'm asking.

**Comment by rrousselGit** [comment]
It is merged, but the issue wasn't closed

**Comment by tvolkert** [comment]
It was rolled back.

**Comment by HerrNiklasRaab** [comment]
It was rolled back?

Here is an Issue somewhat related, unhandled exceptions inside a stream used by a streambuilder don't print the stack trace to the debug console. Why is this the intended behavior?

https://github.com/flutter/flutter/issues/53384#issue-588810501

**Comment by iapicca** [comment]
testing the code [above](https://github.com/flutter/flutter/issues/34545#issue-456555462)
with the latest master 

<details><summary>logs</summary>

```bash
[ +111 ms] Syncing files to device Pixel 3a...
[   +1 ms] Scanning asset files
[   +3 ms] <- reset
[        ] Compiling dart to kernel with 0 updated files
[   +1 ms] <- recompile package:master/main.dart 09a4e4b8-4d74-4a3c-b9d7-c389be6951fa
[        ] <- 09a4e4b8-4d74-4a3c-b9d7-c389be6951fa
[ +102 ms] Updating files
[ +247 ms] DevFS: Sync finished
[   +1 ms] Syncing files to device Pixel 3a... (completed in 357ms)
[   +1 ms] Synced 1.6MB.
[   +2 ms] <- accept
[   +6 ms] Connected to _flutterView/0x7082004820.
[   +3 ms] Flutter run key commands.
[   +3 ms] r Hot reload. 🔥🔥🔥
[   +1 ms] R Hot restart.
[        ] h Repeat this help message.
[        ] d Detach (terminate "flutter run" but leave application running).
[        ] c Clear the screen
[        ] q Quit (terminate the application on the device).
[        ] An Observatory debugger and profiler on Pixel 3a is available at: http://127.0.0.1:56751/koBfi23CAWg=/
[+60093 ms] DevFS: Deleting filesystem on the device (file:///data/user/0/com.example.master/code_cache/masterLXDRLP/master/)
[  +20 ms] DevFS: Deleted filesystem on the device (file:///data/user/0/com.example.master/code_cache/masterLXDRLP/master/)
[ +285 ms] Service protocol connection closed.
[        ] Application finished.
[   +6 ms] executing: /Users/nevercode/Library/Android/sdk/platform-tools/adb -s 965AY0WP5C forward --list
[   +7 ms] Exit code 0 from: /Users/nevercode/Library/Android/sdk/platform-tools/adb -s 965AY0WP5C forward --list
[        ] 965AY0WP5C tcp:56751 tcp:41719
[   +1 ms] executing: /Users/nevercode/Library/Android/sdk/platform-tools/adb -s 965AY0WP5C forward --remove tcp:56751
[  +12 ms] executing: /Users/nevercode/Library/Android/sdk/platform-tools/adb -s 965AY0WP5C forward --list
[  +11 ms] Exit code 0 from: /Users/nevercode/Library/Android/sdk/platform-tools/adb -s 965AY0WP5C forward --list
[   +2 ms] "flutter run" took 105,175ms.
[   +6 ms] Running shutdown hooks
[        ] Shutdown hook priority 4
[   +1 ms] Shutdown hooks complete
[   +1 ms] exiting with code 0

```

</details>


<details><summary>screenshot</summary>

![Screenshot_20200731-140503](https://user-images.githubusercontent.com/45637986/89029457-2bd0fd00-d337-11ea-8269-aaf710d14869.png)

</details>


<details><summary>doctor</summary>

```console
[✓] Flutter (Channel master, 1.21.0-6.0.pre.129, on Mac OS X 10.15.5 19F101, locale en-GB)
    • Flutter version 1.21.0-6.0.pre.129 at /Users/nevercode/development/flutter_master
    • Framework revision 30aef0a3b9 (14 hours ago), 2020-07-30 14:20:21 -0700
    • Engine revision a9910e409c
    • Dart version 2.10.0 (build 2.10.0-1.0.dev 24c7666def)


[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /Users/nevercode/Library/Android/sdk
    • Platform android-29, build-tools 29.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.5)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.5, Build version 11E608c
    • CocoaPods version 1.9.0

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 46.0.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] VS Code
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.12.2

[✓] Connected device (4 available)
    • Pixel 3a (mobile) • 965AY0WP5C • android-arm64  • Android 10 (API 29)
    • macOS (desktop)   • macos      • darwin-x64     • Mac OS X 10.15.5 19F101
    • Web Server (web)  • web-server • web-javascript • Flutter Tools
    • Chrome (web)      • chrome     • web-javascript • Google Chrome 84.0.4147.105

• No issues found!
```

</details>

the issue persists
Thank you

**Comment by giorgio79** [comment]
Going back to the original title of this issue, could we also get stack trace with hasError? I tried 

```
                      if (snapshot.hasError) {
                        throw (snapshot.error);
                        return Text('Error: ${snapshot.error}');
                      }
```
but the stack trace I am getting gives me the line where I call the throw statement as the source of error, not the actual error in the code....

**Comment by awhitford** [comment]
@giorgio79 Don't you need to use `rethrow`?  See [How do I rethrow an exception and preserve the stack trace?](https://stackoverflow.com/questions/16028565/how-do-i-rethrow-an-exception-and-preserve-the-stack-trace).

**Comment by creativecreatorormaybenot** [comment]
@awhitford No, `rethrow` only works in `catch` blocks.

Throwing with a custom (or original) stack trace is not support. See https://github.com/dart-lang/sdk/issues/10297.

**Comment by Manish142** [comment]
I am using StreamProvider to access the data stored on cloud firestore using QuerySnapshot. But I am not able to select the 'intialData' parameter. I am getting the error: "The named parameter 'initialData' is required, but there's no corresponding argument". The few lines of code are as follow: 
 

return StreamProvider<QuerySnapshot>.value(
     
      value: DatabaseService().userdetail,
      child: Scaffold(
        drawer: Drawer(

**Comment by darshankawar** [comment]
Since this issue is open and based on original code sample, there's no logs in the console, I am keeping this issue open and labeling as applicable.

<details>
<summary> stable, master flutter doctor -v </summary>

```
[✓] Flutter (Channel stable, 3.3.10, on macOS 12.2.1 21D62 darwin-x64, locale
    en-GB)
    • Flutter version 3.3.10 on channel stable at
      /Users/dhs/documents/fluttersdk/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 135454af32 (15 hours ago), 2022-12-15 07:36:55 -0800
    • Engine revision 3316dd8728
    • Dart version 2.18.6
    • DevTools version 2.15.0

[!] Xcode - develop for iOS and macOS (Xcode 12.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    ! Flutter recommends a minimum Xcode version of 13.
      Download the latest version or update via the Mac App Store.
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] VS Code (version 1.62.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[✓] Connected device (5 available)
    • SM G975F (mobile)       • RZ8M802WY0X • android-arm64   • Android 11 (API 30)
    • Darshan's iphone (mobile)  • 21150b119064aecc249dfcfe05e259197461ce23 •
      ios            • iOS 14.4.1 18D61
    • iPhone 12 Pro Max (mobile) • A5473606-0213-4FD8-BA16-553433949729     •
      ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • macOS (desktop)            • macos                                    •
      darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
    • Chrome (web)               • chrome                                   •
      web-javascript • Google Chrome 98.0.4758.80

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.

[!] Flutter (Channel master, 3.7.0-17.0.pre.21, on macOS 12.2.1 21D62
    darwin-x64, locale en-GB)
    • Flutter version 3.7.0-17.0.pre.21 on channel master at
      /Users/dhs/documents/fluttersdk/flutter
    ! Warning: `flutter` on your path resolves to
      /Users/dhs/Documents/Fluttersdk/flutter/bin/flutter, which is not inside
      your current Flutter SDK checkout at
      /Users/dhs/documents/fluttersdk/flutter. Consider adding
      /Users/dhs/documents/fluttersdk/flutter/bin to the front of your path.
    ! Warning: `dart` on your path resolves to
      /Users/dhs/Documents/Fluttersdk/flutter/bin/dart, which is not inside your
      current Flutter SDK checkout at /Users/dhs/documents/fluttersdk/flutter.
      Consider adding /Users/dhs/documents/fluttersdk/flutter/bin to the front
      of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 55e3894a8c (2 hours ago), 2023-01-10 21:37:34 -0500
    • Engine revision 53bd4bbf96
    • Dart version 3.0.0 (build 3.0.0-108.0.dev)
    • DevTools version 2.20.0
    • If those were intentional, you can disregard the above warnings; however
      it is recommended to use "git" directly to perform update checks and
      upgrades.

[!] Xcode - develop for iOS and macOS (Xcode 12.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    ! Flutter recommends a minimum Xcode version of 13.
      Download the latest version or update via the Mac App Store.
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] VS Code (version 1.62.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[✓] Connected device (5 available)
    • SM G975F (mobile)       • RZ8M802WY0X • android-arm64   • Android 11 (API 30)
    • Darshan's iphone (mobile)  • 21150b119064aecc249dfcfe05e259197461ce23 •
      ios            • iOS 14.4.1 18D61
    • iPhone 12 Pro Max (mobile) • A5473606-0213-4FD8-BA16-553433949729     •
      ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • macOS (desktop)            • macos                                    •
      darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
    • Chrome (web)               • chrome                                   •
      web-javascript • Google Chrome 98.0.4758.80

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.



```

</details>
